jobs/sentry-error-loader/sentry_error_loader.py

```diff
@@ -0,0 +1,111 @@
+import gzip
+import os
+from typing import ClassVar, List, Optional
+
+import clickhouse_connect  # type: ignore
+import pandas as pd
+import pendulum
+import typer
+from calitp_data_infra.storage import (  # type: ignore
+    PartitionedGCSArtifact,
+    get_fs,
+    make_name_bq_safe,
+)
+
+CALITP_BUCKET__SENTRY_EVENTS = os.environ["CALITP_BUCKET__SENTRY_EVENTS"]
+
+
+def process_arrays_for_nulls(arr):
+    """
+    BigQuery doesn't allow arrays that contain null values --
+    see: https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#array_nulls
+    Therefore we need to manually replace nulls with falsy values according
+    to the type of data we find in the array.
+    """
+    types = set(type(entry) for entry in arr if entry is not None)
+
+    if not types:
+        return []
+    # use empty string for all non-numeric types
+    # may need to expand this over time
+    filler = -1 if types <= {int, float} else ""
+    return [x if x is not pd.NA else filler for x in arr]
+
+
+def fetch_and_clean_from_clickhouse(project_slug, target_date):
+    """
+    Download Sentry event records from Clickhouse as a DataFrame.
+    """
+
+    print(f"Gathering Sentry event data for project {project_slug}")
+    client = clickhouse_connect.get_client(host="localhost", port=8123)
+    all_rows = client.query_df(
+        f"""SELECT * FROM errors_dist
+                                   WHERE message like '%RTFetchException%'
+                                   and toDate(timestamp) == '{target_date}'"""
+    )
+
+    cleaned_df = all_rows.rename(make_name_bq_safe, axis="columns")
+
+    cols_with_nulls_in_arrays = ["exception_frames_colno", "exception_frames_package"]
+    for col_name in cols_with_nulls_in_arrays:
+        cleaned_df[col_name] = cleaned_df[col_name].apply(
+            lambda row: process_arrays_for_nulls(row)
+        )
+
+    """
+    Pandas' default timestamp datatype returned by clickhouse-connect writes out unix timestamps
+    in nanoseconds, but they get interpreted by BQ as microseconds. A quick cast to datetime string
+    before writing avoids triggering the issue.
+    """
+    cols_with_unix_timestamps = ["timestamp", "message_timestamp", "received"]
+    for col_name in cols_with_unix_timestamps:
+        cleaned_df[col_name] = cleaned_df[col_name].apply(lambda row: str(row))
```

Comment:
I think you can use `astype()` here

Reply:
Resolved by [d33e6dc](https://github.com/cal-itp/data-infra/pull/2345/commits/d33e6dcb178194c5bd130c5f20abfdb340015fdc)